File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # check horizontal
        for i in range(9):
            row = set()
            for j in range(9):
                val = board[i][j]
                if val == ".":
                    continue
                if val in row:
                    return False
                row.add(val)
        
        # check vertical
        for i in range(9):
            col = set()
            for j in range(9):
                val = board[j][i]
                if val == ".":
                    continue
                if val in col: 
                    return False
                else:
                    col.add(val)

        # check square
        square = [[set() for _ in range(3)] for _ in range(3)]
        for i in range(9):
            for j in range(9):
                val = board[i][j]
                if val == ".":
                    continue
                if val in square[i//3][j//3]:
                    return False
                else:
                    square[i//3][j//3].add(board[i][j])

        return True

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int i=0; i<9; i++){
            Set<Character> set = new HashSet<>();
            for (int j=0; j<9; j++){
                char value = board[i][j];
                if (value == '.') continue;
                
                if (set.contains(value)){
                    return false;
                };
                set.add(value);
            }
        }

        for (int i=0; i<9; i++){
            Set<Character> set = new HashSet<>();
            for (int j=0; j<9; j++){
                char value = board[j][i];
                if (value == '.') continue;

                if (set.contains(value)){
                    return false;
                }
                set.add(value);
            }
        }

        for (int i=0; i<9; i++){
            Set<Character> set = new HashSet<>();
            int row = (i/3) * 3;
            int col = (i%3) * 3;

            for (int j=0; j<3; j++){
                for (int k=0; k<3; k++){
                    char value = board[row+j][col+k];
                    if (value=='.') continue;
                    if (set.contains(value)){
                        return false;
                    }
                    set.add(value);
                }
            }
        }

        return true;
    }
}